Add analyser that verifies suspicious angles in railway

Tristramg opened the pull request:
This analyser is quite similar to `Analyser_Osmosis_Way_Approximate` however it works a bit differently as it verifies the angle consecutive nodes of a way. It can make sens on roads, but very rarely on trains.

A few things I’m still not sure:
- I’m still confused about the class and item I should use
- Translations: I kept pretty much the same wording as the Way Approximate analyser as the analyses still make sens, but I didn’t how I could generate the right entries for the `fr.po`

Famlam commented:
Just for my understanding, how do these results differ with the results from the existing analyser, e.g. http://osmose.openstreetmap.fr/nl/issues/open?source=33602&item=1190&class=10 ? (Except for that it includes light rail)?

Because if there's a lot of duplicate reports, with basically the same message, it may be relevant to UNION the results? Or exclude those ways which have been detected already? (Just my opinion)

Don't worry about the translations, that's synchronized :)

P.s. sharp angles are wrong in any rail, not just rail and light_rail, right? Or does this give too much output?

Famlam commented:
> From my understating from the other analyser, it tries to compute radiuses instead of angles. Both are complementary and won’t detect the same things. (But my trigonometry is bad, maybe I missed an important point).

I can run a comparison to see how much overlap there is on some countries, but won't be able to do so until mid/end next week. I suspect that the bad radiuses will commonly be correlated with bad angles, but not sure.

> Other (significant) railway are narrow_gauge, subway and tram. Because they are used with smaller vehicles, sharper angles are more common. Maybe an approach would be to have different angles depending on the kind of rail, but that would require some trial and error to find good values.

Yes, and if the angle becomes a variable, it may be implemented for i.e. (car) highways too. A 170° turn in one node sounds unlikely to be precise. But this would be a next PR I think

frodrigo commented:
`Analyser_Osmosis_Way_Approximate` look for circle radius. While angle depend on map data scale, radius not. With a low resolution of highway or railway, a shape angle could be see but at hight resolution it will be not so sharp for the same radius. It is all about level of detail in OSM data.

Maybe computing direct angle make sens with small way segment only, so we sure we are at hight resolution?

Tristramg commented:
Indeed, the distance make a difference. What we were trying to catch are
more likely to be mapping errors, like a node dragged by accident.
Maybe restricting this analyser when both segments are less than 100m might
make sense? What do you think?

If yes, the projection passed as parameters computes distances as meters,
am I right?

On Thu, 10 Aug 2023 at 11:05, Frédéric Rodrigo ***@***.***>
wrote:

> Analyser_Osmosis_Way_Approximate look for circle radius. While angle
> depend on map data scale, radius not. With a low resolution of highway or
> railway, a shape angle could be see but at hight resolution it will be not
> so sharp for the same radius. It is all about level of detail in OSM data.
>
> Maybe computing direct angle make sens with small way segment only, so we
> sure we are at hight resolution?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/osm-fr/osmose-backend/pull/1983#issuecomment-1672842126>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAC7SZYZ4XLFUQJBEAYIZTXUSP6NANCNFSM6AAAAAA3IYEJZQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

frodrigo commented:
> Indeed, the distance make a difference. What we were trying to catch are
more likely to be mapping errors, like a node dragged by accident.
Maybe restricting this analyser when both segments are less than 100m might
make sense? What do you think?

I think both are interesting

> If yes, the projection passed as parameters computes distances as meters,
am I right?

yes

Tristramg commented:
I added a length limit, a comment to explain the difference with the existing analyser and changed a bit the explanation strings.

An example could be https://www.openstreetmap.org/node/4983237612#map=21/48.21104/12.58973&layers=T
![image](https://github.com/osm-fr/osmose-backend/assets/12235/fded37e3-4a7b-4690-9984-70fe4103dbca)

Famlam commented on analysers/analyser_osmosis_railway_angle.py:
```diff
@@ -0,0 +1,105 @@
+                     lead(geom, 2) OVER (PARTITION BY id ORDER BY index)
+                )
+            ) angle,
+            ST_Length(geom, lead(geom, 1) OVER (PARTITION BY id ORDER BY index)) AS length_1,
```
2023-08-16 10:31:29     ST_Length(geom, lead(geom, 1) OVER (PARTITION BY...
2023-08-16 10:31:29     ^
2023-08-16 10:31:29     HINT:  No function matches the given name and argument types. You might need to add explicit type casts.

Tristramg replied:
Oh sorry :facepalm: 
I pushed the fixed version

Famlam commented:
Comparing the results of the analysers I guess there's no need to filter to prevent duplicates: most results are unique.

| country | # way approx | (of which) # railway approx | # suspicious angle | # of duplicate reports |
| --- | --- | --- | --- | --- |
| venezuela | 11796 | 1 | 0 | 0 |
| usa_iowa | 134 | 3 | 13 | 0 |
| slovenia | 22 | 0 | 9 | 0 |
| japan_chubu | 127 | 0 | 27 | 0 |
| kazakhstan | 19222 (2727 ways) | 689 | 164 (158 ways) | 2 ways |
| china_xinjiang | 15248 | 246 | 19 | 0 |

Results attached: [results.zip](https://github.com/osm-fr/osmose-backend/files/12360905/angle.zip)

Famlam commented on analysers/analyser_osmosis_railway_angle.py:
```diff
@@ -0,0 +1,111 @@
+
+`railway=rail` crudely drawn.'''))
+
+        self.tags = ("rail", "lightrail")
```
https://wiki.openstreetmap.org/wiki/Tag:railway%3Dlight_rail
light_rail with an underscore

Famlam commented on analysers/analyser_osmosis_railway_angle.py:
```diff
@@ -0,0 +1,111 @@
+`railway=rail` crudely drawn.'''))
+
+        self.tags = ("rail", "lightrail")
+        self.callback10 = lambda res: {"class":10, "subclass":res[2], "data":[self.way_full, self.positionAsText], "text": T_("railway with suspicious angle {0}°", res[1])}
```
Analyser way_approximate uses the same class and item id. To allow users to filter, I would use class `11` here (and thus also in line 90)

---

Additionally, I think the angle is in `res[2]`?
The current message reads: `railway with suspicious angle POINT(81.36814960000001 50.936915400000004)°` with `subclass="352.6731599505589"` (the angle).
(I think subclass must be an integer? Maybe this is where you wanted to use `index` (=`res[4]`)?

`res[3]` (the value of tag `railway=*`) is currently not used.

Famlam commented on analysers/analyser_osmosis_railway_angle.py:
```diff
@@ -0,0 +1,111 @@
+            fix = T_(
+'''After checking orthophotos, add nodes or move existing nodes.'''),
+            trap = T_(
+'''On service ways, train stations, train workshops that may be either a
```
This is also in the other analyser and should probably be fixed in both, but I think the grammar of this line could be improved. Suggestion:
`On service ways, train stations and train workshops this could be a false positive.`

Famlam commented on analysers/analyser_osmosis_railway_angle.py:
```diff
@@ -0,0 +1,111 @@
+        self.classs_change[10] = self.def_class(item = 1190, level = 2, tags = ['geom', 'railway', 'fix:imagery'],
+            title = T_('Suspicious angle in railway'),
+            detail = T_(
+'''Sharp angles on a railway are suspicious. Maybe a node was accidentaly dragged'''),
```
Typo: `accidentally`
Also, please end the line with a `.` or `?` ;)

Famlam commented on analysers/analyser_osmosis_railway_angle.py:
```diff
@@ -0,0 +1,111 @@
+        AND length_2 < 100
+"""
+
+class Analyser_Osmosis_Rail_Suspicious_Angle(Analyser_Osmosis):
```
Just with an eye on the future: 
I guess this analyser (with flexible angle restrictions) could in the future be expanded to include car highways with sharp angles. Maybe better to name it more neutral with `way` instead of `railway`. (Also, I guess it's convenient to have the same file name as analyser class name)

Tristramg replied:
I changed it to way, but I not yet convinced that it would make sens for roads and might result in a lot of false positive (e.g. https://www.openstreetmap.org/way/689270247)

Famlam replied:
Yes, for roads (for cars) it should probably be limited to e.g. <30° angles or so, definitely less than 90 or even 150°.

Problem for the future, not for now 😉

Famlam commented on analysers/analyser_osmosis_railway_angle.py:
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python
```
General note: to enable the analyser for all countries, it must be added to this list too:
https://github.com/osm-fr/osmose-backend/blob/dev/osmose_config.py#L197

Tristramg commented:
Thank you for this precious review!

Famlam reviewed (approved):
No further comments from me :)
p.s. ignore the "Famlam self-requested a review". Wrong button, I wanted to remove my earlier "requested changes" review ;)

frodrigo commented on analysers/analyser_osmosis_way_angle.py:
```diff
@@ -0,0 +1,110 @@
+        SELECT
+            id, index, type,
+            lead(geom4326) OVER (PARTITION BY id ORDER BY index) AS geom4326,
+            lead(geom) OVER (PARTITION BY id ORDER BY index) AS geom,
```
Could be removed.

```suggestion
            lead(geom) OVER (PARTITION BY id ORDER BY index) AS geom,
```

Tristramg replied:
I’m don’t understand the suggested change, it looks the same. What could be removed?

Famlam replied:
The fix suggestion is strange indeed, but I think @frodrigo means that `geom` isn't used, i.e. line 52 can be removed

Tristramg replied:
Thank you! It’s removed

frodrigo commented:
Merged. Thank you all.

frodrigo commented:
The result is live
https://osmose.openstreetmap.fr/en/issues/open?item=1190&class=11

Tristramg commented:
Thanks both of you for your patience to get me to a working plugin!
